indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java

```diff
@@ -225,6 +270,79 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
     return workerParallelIndexCapacity;
   }
 
+  private boolean hasSufficientWorkerCapacity(Task task)
+  {
+    int capacityRemaining = worker.getCapacity() - getCurrCapacityUsed();
+    int requiredCapacity = task.getTaskResource().getRequiredCapacity();
+    return capacityRemaining >= requiredCapacity;
+  }
+
+  private boolean isAvailabilityGroupAvailable(Task task)
+  {
+    return !getAvailabilityGroups().contains(task.getTaskResource().getAvailabilityGroup());
+  }
+
+  private boolean canRunTaskBasedOnCustomLimit(Task task, Map<String, Integer> limitsMap, Map<String, Double> ratiosMap)
+  {
+    final Integer limit = getLimitForTask(task.getType(), limitsMap, ratiosMap);
+
+    if (limit == null) {
+      return true; // No limit specified, so task can run
+    }
+
+    int currentCapacityUsed = getCurrCapacityUsedByTaskType().getOrDefault(task.getType(), 0);
+    int requiredCapacity = task.getTaskResource().getRequiredCapacity();
+
+    return hasCapacityBasedOnLimit(limit, currentCapacityUsed, requiredCapacity);
+  }
+
+  private Integer getLimitForTask(
+      String taskType,
+      Map<String, Integer> limitsMap,
+      Map<String, Double> ratiosMap
+  )
+  {
+    Integer absoluteLimit = limitsMap.get(taskType);
+    Double ratioLimit = ratiosMap.get(taskType);
+
+    if (absoluteLimit == null && ratioLimit == null) {
+      return null;
+    }
+
+    // Validate the absolute limit if present
+    if (absoluteLimit != null) {
+      Preconditions.checkArgument(absoluteLimit >= 0, "Absolute limit for task %s must be non-negative.", taskType);
+    }
+
+    // Validate the ratio limit if present
+    if (ratioLimit != null) {
+      Preconditions.checkArgument(ratioLimit >= 0.0 && ratioLimit <= 1.0,
+                                  "Ratio for task %s must be between 0.0 and 1.0 inclusive.", taskType
+      );
+    }
+
+    final int totalCapacity = worker.getCapacity();
+
+    Integer ratioBasedLimit = ratioLimit != null ? calculateTaskCapacityFromRatio(ratioLimit, totalCapacity) : null;
+
+    if (absoluteLimit != null && ratioBasedLimit != null) {
+      return Math.min(absoluteLimit, ratioBasedLimit);
+    }
+
+    return absoluteLimit != null ? absoluteLimit : ratioBasedLimit;
+  }
+
+  private boolean hasCapacityBasedOnLimit(int limit, int currentCapacityUsed, int requiredCapacity)
```

Comment:
This method is not needed, it can be inlined.

Reply:
Done, thanks for notice